File: week01/pset1/credit/credit.c
// Calculate the validity of credit card number
// Step: get number, count digit, odd digit--add, even digit 0-4 multiply, 5-9 2i+1
#include <cs50.h>
#include <stdio.h>

long get_card();
void decide_company(int sum, int count, int left1, int left2);

int main(void)
{
    // Get positive input
    long card_num = get_card();

    // Calculate digits and luhn sum
    int digit_cnt = 0;
    int sum = 0, remainder = 0;

    // left1 for most left digit, left2 for second left digit
    int left1 = 0, left2 = 0;

    // Get each digit from right to left, save the most 2 left digit to decide company
    while (card_num > 0)
    {
        digit_cnt++;
        left2 = left1;
        remainder = card_num % 10;
        card_num /= 10;
        left1 = remainder;
        if (digit_cnt % 2 != 0)
        {
            sum += remainder;
        }
        else
        {
            if (remainder < 5)
            {
                sum += remainder * 2;
            }
            else
            {
                sum += 1 + (remainder - 5) * 2;
            }
        }
    }

    // If luhn sum valid, decide company by digits
    decide_company(sum, digit_cnt, left1, left2);
}

long get_card()
{
    long card_num = 0;
    do
    {
        card_num = get_long("Number: ");
    }
    while (card_num < 0);
    return card_num;
}

// Given luhn sum, card digits. Decide valid or not
void decide_company(int sum, int count, int left1, int left2)
{
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        if (left1 == 4 && (count == 13 || count == 16))
        {
            printf("VISA\n");
        }
        else if (left1 == 5 && left2 < 6 && left2 > 0 && count == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (left1 == 3 && (left2 == 4 || left2 == 7) && count == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

File: week06/pset6/sentimental-credit/credit.py
card = input("Number: ")
bits = len(card)

flip = 0
digit_sum = 0
for i in reversed(card):
    if flip == 0:
        flip = 1
        digit_sum += int(i) - 0
    else:
        flip = 0
        even = int(i) - 0
        if even <5:
            even *= 2
        else:
            even = 1 + 2 * (even - 5)
        digit_sum += even

if (digit_sum % 10) == 0:
    if bits in [13, 16] and card[0] == '4':
        print("VISA")
    elif (bits == 15) and (card[0:2] in ['34', '37']):
        print("AMEX")
    elif (bits == 16) and (int(card[0:2]) in range(51, 56)):
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

